src/mon/OSDMonitor.cc

```diff
@@ -10436,13 +10437,19 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
       }
 
       for (auto &osd : osds) {
-        if (prefix == "osd down") {
+        if (prefix == "osd down" ||
+            prefix == "osd stop") {
 	  if (osdmap.is_down(osd)) {
             if (verbose)
 	      ss << "osd." << osd << " is already down. ";
 	  } else {
             pending_inc.pending_osd_state_set(osd, CEPH_OSD_UP);
-	    ss << "marked down osd." << osd << ". ";
+            if (prefix == "osd stop") {
+              pending_inc.pending_osd_state_set(osd, CEPH_OSD_STOP);
+              ss << "will stop osd." << osd << ". ";
+            } else {
+	      ss << "marked down osd." << osd << ". ";
+            }
```

Comment:
one thing here: if you run ``ceph osd stop`` when the osd is running but marked down, we won't set the flag, and the daemon won't stop.  I think we should separate out the down and stop cases, but make the stop case set both CEPH_OSD_OP|CEPH_OSD_DOWN.